src/main/java/io/cdap/plugin/gcp/spanner/connector/SpannerConnector.java

```diff
@@ -0,0 +1,276 @@
+    int countLimit = limit == null || limit <= 0 ? Integer.MAX_VALUE : limit;
+    int count = 0;
+    BrowseDetail.Builder browseDetailBuilder = BrowseDetail.builder();
+    String pathPrefix = "/" + instance + "/";
+    for (Database database : page.iterateAll()) {
+      if (count < countLimit) {
+        String name = database.getId().getDatabase();
+        browseDetailBuilder
+          .addEntity(BrowseEntity.builder(name, pathPrefix + name, ENTITY_TYPE_DATABASE).canBrowse(true).build());
+      }
+      count++;
+    }
+    return browseDetailBuilder.setTotalCount(countLimit).build();
+  }
+
+  private BrowseDetail listTables(Spanner spanner, String instance, String database, Integer limit) {
+    ResultSet resultSet = spanner.getDatabaseClient(DatabaseId.of(config.getProject(), instance, database)).
+      singleUse().executeQuery(LIST_TABLES_STATEMENT);
+    int countLimit = limit == null || limit <= 0 ? Integer.MAX_VALUE : limit;
+    int count = 0;
+    BrowseDetail.Builder browseDetailBuilder = BrowseDetail.builder();
+    String pathPrefix = "/" + instance + "/" + database + "/";
+    while (resultSet.next()) {
+      String name = resultSet.getString("table_name");
+      browseDetailBuilder
+        .addEntity(BrowseEntity.builder(name, pathPrefix + name, ENTITY_TYPE_TABLE).canSample(true).build());
+    }
+    return browseDetailBuilder.setTotalCount(countLimit).build();
+  }
+
+  private BrowseDetail getTableDetail(Spanner spanner, String instance, String database, String table) {
+
+    Statement getTableSchemaStatement = SCHEMA_STATEMENT_BUILDER.bind(TABLE_NAME).to(table).build();
+    try (ResultSet resultSet = spanner.getDatabaseClient(DatabaseId.of(config.getProject(), instance, database)).
+      singleUse().executeQuery(getTableSchemaStatement)) {
+      if (resultSet.next()) {
+        String path = "/" + instance + "/" + database + "/" + table;
+        return BrowseDetail.builder()
+          .addEntity(BrowseEntity.builder(table, path, ENTITY_TYPE_TABLE).canSample(true).build())
+          .setTotalCount(1)
+          .build();
+      } else {
+        throw new IllegalArgumentException(String.format("Cannot find table: %s", table));
+      }
+    }
+  }
+
+  private List<StructuredRecord> getTableData(String instance, String database, String table, int limit)
+    throws IOException {
+    List<StructuredRecord> records = new ArrayList<>();
+    try (Spanner spanner = getSpanner()) {
+      Schema schema = getTableSchema(instance, database, table);
+      List<String> columnNames = schema.getFields().stream().map(Schema.Field::getName).collect(Collectors.toList());
+      ResultSet resultSet = spanner.getDatabaseClient(DatabaseId.of(config.getProject(), instance, database))
+        .singleUse()
+        .read(table, KeySet.all(), columnNames, Options.limit(limit));
+      while (resultSet.next()) {
+        records.add(SpannerConnectorUtil.parseStructuredRecordFromResultSet(schema, resultSet));
```

Comment:
noticed we already have https://github.com/data-integrations/google-cloud/blob/develop/src/main/java/io/cdap/plugin/gcp/spanner/source/ResultSetToRecordTransformer.java, can we just reuse that?

Reply:
thanks for pointing that out, fixed